contributions/views.py

```diff
@@ -1,158 +1,266 @@
                 schedule=schedule,
+                member=membership,
                 amount=amount,
-                method='MPESA',
-                reference=receipt,
-                status='SUCCESS'
+                method=Contribution.PaymentMethod.CASH,
+                status=status_val
             )
 
-            chama.balance += contribution.amount
-            chama.save()
+        output = ContributionSerializer(contrib, context={'request': request}).data
+        return Response(output, status=status.HTTP_201_CREATED)
 
-            ActivityLog.objects.create(
-                user=user,
-                chama=chama,
-                action='CONTRIBUTION',
-                details=f"M-Pesa contribution of KES {float(amount):.2f}"
+    @action(detail=True, methods=['patch'], url_path='update-status')
+    def update_status(self, request, pk=None):
+        """
+        Only chama ADMIN/TREASURER may change status of PENDING contributions.
+        """
+        contrib = self.get_object()
+        user    = request.user
+
+        # verify admin/treasurer role
+        if not Membership.objects.filter(
+            user=user,
+            chama=contrib.schedule.chama,
+            role__in=[Membership.Role.ADMIN, Membership.Role.TREASURER]
+        ).exists():
+            raise PermissionDenied("Only chama admins or treasurers can update status.")
+
+        # only pending may change
+        if contrib.status != Contribution.Status.PENDING:
+            raise ValidationError({'status': 'Only pending contributions can be updated.'})
+
+        serializer = self.get_serializer(contrib, data=request.data, partial=True)
+        serializer.is_valid(raise_exception=True)
+        new_status = serializer.validated_data['status']
+
+        with transaction.atomic():
+            serializer.save()
+            # bump balance on approval
+            if new_status == Contribution.Status.APPROVED:
+                chama = contrib.schedule.chama
+                chama.balance = F('balance') + contrib.amount
+                chama.save(update_fields=['balance'])
+
+        return Response({'detail': 'Status updated successfully.'}, status=status.HTTP_200_OK)
+
+    @action(detail=False, methods=['post'], url_path='initiate-mpesa')
+    def initiate_mpesa(self, request):
+        """
+        Kick off an STK Push for a member’s scheduled contribution.
+        """
+        user = request.user
+        schedule_id = request.data.get('schedule')
+        amount      = request.data.get('amount')
+
+        # Validate schedule & membership
+        try:
+            schedule_uuid = uuid.UUID(schedule_id)
+            schedule = ContributionSchedule.objects.get(id=schedule_uuid)
+        except (ValueError, DjangoValidationError, ContributionSchedule.DoesNotExist):
+            raise ValidationError({'schedule': 'Invalid schedule ID.'})
+        except Membership.DoesNotExist:
+            raise PermissionDenied('Not a member of this Chama.')
```

Comment:
This block is unnecessary here since there's no membership lookup in the try block. You can remove it.